File: backend/src/payroll_copilot/application/services/payslip_field_registry.py
"""Payslip Field Registry — presentation metadata over existing canonical keys.

KEEP AS-IS:
  - DynamicDocumentEntry (extraction SoT)
  - PAYSLIP_FIELD_KEYS / resolve_canonical_key / projector behavior

This registry is additive metadata for Digital Payslip projection.
It does NOT change extraction, confirmation gates, or persistence blockers.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
from typing import Iterable

from payroll_copilot.application.ports.payslip_parser import (
    PAYSLIP_CANONICAL_EXTRA_KEYS,
    PAYSLIP_FIELD_KEYS,
)
from payroll_copilot.application.services.dynamic_document import resolve_canonical_key


class FieldRequirementCategory(StrEnum):
    REQUIRED = "required"
    EXPECTED = "expected"
    OTHER = "other"


class PayslipFieldSection(StrEnum):
    IDENTITY = "identity"
    EMPLOYER = "employer"
    PERIOD = "period"
    EARNINGS = "earnings"
    DEDUCTIONS = "deductions"
    PAYMENT = "payment"
    OTHER = "other"


@dataclass(frozen=True, slots=True)
class PayslipFieldDefinition:
    """Known payslip field presentation definition."""

    canonical_key: str
    label_i18n_key: str
    section: PayslipFieldSection
    display_order: int
    requirement_category: FieldRequirementCategory
    editable: bool = True
    # Reserved for future policy — NOT enforced as a new persistence gate.
    required_on_payslip: bool = False
    required_for_persistence: bool = False


# National ID is the government identity field (required_on_payslip).
# employee_id is the payroll/system employee identifier (expected) — never the same concept.
# employee_number remains the organization employee number.
_REQUIRED_KEYS: tuple[tuple[str, PayslipFieldSection, int], ...] = (
    ("employee_name", PayslipFieldSection.IDENTITY, 10),
    ("national_id", PayslipFieldSection.IDENTITY, 20),
    ("employer_name", PayslipFieldSection.EMPLOYER, 25),
    ("employer_id", PayslipFieldSection.EMPLOYER, 26),
    ("employer_address", PayslipFieldSection.EMPLOYER, 27),
    ("employment_start_date", PayslipFieldSection.IDENTITY, 28),
    ("employment_scope", PayslipFieldSection.IDENTITY, 29),
    ("pay_period", PayslipFieldSection.PERIOD, 30),
    ("base_salary", PayslipFieldSection.EARNINGS, 40),
    ("salary_calculation_basis", PayslipFieldSection.EARNINGS, 45),
    ("gross_salary", PayslipFieldSection.EARNINGS, 50),
    ("income_tax", PayslipFieldSection.DEDUCTIONS, 60),
    ("national_insurance", PayslipFieldSection.DEDUCTIONS, 70),
    ("total_deductions", PayslipFieldSection.DEDUCTIONS, 80),
    ("net_salary", PayslipFieldSection.PAYMENT, 90),
    ("amount_paid", PayslipFieldSection.PAYMENT, 95),
    ("payment_method", PayslipFieldSection.PAYMENT, 100),
    ("minimum_wage_monthly", PayslipFieldSection.OTHER, 105),
    ("minimum_wage_hourly", PayslipFieldSection.OTHER, 106),
)

_EXPECTED_KEYS: tuple[tuple[str, PayslipFieldSection, int], ...] = (
    ("employee_number", PayslipFieldSection.IDENTITY, 110),
    ("employee_id", PayslipFieldSection.IDENTITY, 115),  # payroll/system ID — not National ID
    ("employment_type", PayslipFieldSection.IDENTITY, 120),
    ("seniority_years", PayslipFieldSection.IDENTITY, 125),
    ("department", PayslipFieldSection.IDENTITY, 130),
    ("hourly_rate", PayslipFieldSection.EARNINGS, 140),
    ("regular_hours", PayslipFieldSection.EARNINGS, 150),
    ("overtime_hours", PayslipFieldSection.EARNINGS, 160),
    ("travel_expenses", PayslipFieldSection.EARNINGS, 170),
    ("health_tax", PayslipFieldSection.DEDUCTIONS, 180),
    ("pension_employee", PayslipFieldSection.DEDUCTIONS, 190),
    ("pension_employer", PayslipFieldSection.DEDUCTIONS, 200),
    ("severance", PayslipFieldSection.DEDUCTIONS, 210),
    ("training_fund", PayslipFieldSection.DEDUCTIONS, 220),
    ("bank_name", PayslipFieldSection.PAYMENT, 230),
    ("bank_branch", PayslipFieldSection.PAYMENT, 231),
    ("bank_account", PayslipFieldSection.PAYMENT, 232),
    ("vacation_balance", PayslipFieldSection.OTHER, 240),
    ("sick_leave_balance", PayslipFieldSection.OTHER, 250),
    ("messages", PayslipFieldSection.OTHER, 260),
)

# Persistence-critical candidates (metadata only — gates already enforce NID/period).
_PERSISTENCE_CRITICAL: frozenset[str] = frozenset(
    {
        "employee_name",
        "national_id",
        "employee_number",
        "pay_period",
    }
)


def _build_registry() -> dict[str, PayslipFieldDefinition]:
    defs: dict[str, PayslipFieldDefinition] = {}
    for key, section, order in _REQUIRED_KEYS:
        defs[key] = PayslipFieldDefinition(
            canonical_key=key,
            label_i18n_key=f"payroll.fields.{key}",
            section=section,
            display_order=order,
            requirement_category=FieldRequirementCategory.REQUIRED,
            editable=True,
            required_on_payslip=True,
            required_for_persistence=key in _PERSISTENCE_CRITICAL,
        )
    for key, section, order in _EXPECTED_KEYS:
        defs[key] = PayslipFieldDefinition(
            canonical_key=key,
            label_i18n_key=f"payroll.fields.{key}",
            section=section,
            display_order=order,
            requirement_category=FieldRequirementCategory.EXPECTED,
            editable=True,
            required_on_payslip=False,
            required_for_persistence=key in _PERSISTENCE_CRITICAL,
        )
    order_cursor = 900
    for key in (*PAYSLIP_FIELD_KEYS, *PAYSLIP_CANONICAL_EXTRA_KEYS):
        if key in defs:
            continue
        defs[key] = PayslipFieldDefinition(
            canonical_key=key,
            label_i18n_key=f"payroll.fields.{key}",
            section=PayslipFieldSection.OTHER,
            display_order=order_cursor,
            requirement_category=FieldRequirementCategory.OTHER,
            editable=True,
            required_on_payslip=False,
            required_for_persistence=False,
        )
        order_cursor += 10
    return defs


_REGISTRY: dict[str, PayslipFieldDefinition] = _build_registry()


def get_field_definition(key: str) -> PayslipFieldDefinition | None:
    normalized = (key or "").strip()
    if not normalized:
        return None
    if normalized in _REGISTRY:
        return _REGISTRY[normalized]
    canonical = resolve_canonical_key(normalized)
    if canonical and canonical in _REGISTRY:
        return _REGISTRY[canonical]
    return None


def list_field_definitions(
    *,
    categories: Iterable[FieldRequirementCategory] | None = None,
) -> list[PayslipFieldDefinition]:
    items = list(_REGISTRY.values())
    if categories is not None:
        allowed = set(categories)
        items = [item for item in items if item.requirement_category in allowed]
    return sorted(items, key=lambda item: (item.display_order, item.canonical_key))


def required_on_payslip_keys() -> frozenset[str]:
    return frozenset(
        item.canonical_key for item in _REGISTRY.values() if item.required_on_payslip
    )


def requirement_category_for_key(key: str) -> FieldRequirementCategory:
    definition = get_field_definition(key)
    if definition is None:
        return FieldRequirementCategory.OTHER
    return definition.requirement_category


def registry_snapshot_for_tests() -> dict[str, dict[str, object]]:
    """Stable dump for FE/BE sync tests."""
    return {
        key: {
            "canonical_key": item.canonical_key,
            "requirement_category": item.requirement_category.value,
            "section": item.section.value,
            "display_order": item.display_order,
            "required_on_payslip": item.required_on_payslip,
            "required_for_persistence": item.required_for_persistence,
            "editable": item.editable,
            "label_i18n_key": item.label_i18n_key,
        }
        for key, item in sorted(_REGISTRY.items())
    }

File: frontend/src/lib/employee/payslip-field-registry.ts
/**
 * Frontend mirror of backend payslip_field_registry.
 * Keep in sync with:
 *   backend/.../application/services/payslip_field_registry.py
 *
 * Presentation metadata only — does not change extraction or persistence gates.
 */

export type FieldRequirementCategory = 'required' | 'expected' | 'other';

export type PayslipFieldSection =
  | 'identity'
  | 'employer'
  | 'period'
  | 'earnings'
  | 'deductions'
  | 'payment'
  | 'other';

export type PayslipFieldDefinition = {
  canonical_key: string;
  label_i18n_key: string;
  section: PayslipFieldSection;
  display_order: number;
  requirement_category: FieldRequirementCategory;
  editable: boolean;
  /** Presentation: show empty when missing. Not a persistence blocker. */
  required_on_payslip: boolean;
  /** Reserved metadata — existing gates remain the persistence SoT. */
  required_for_persistence: boolean;
};

type DefInput = {
  key: string;
  section: PayslipFieldSection;
  order: number;
  category: FieldRequirementCategory;
  required_on_payslip?: boolean;
  required_for_persistence?: boolean;
};

const PERSISTENCE_CRITICAL = new Set([
  'employee_name',
  'national_id',
  'employee_number',
  'pay_period',
]);

const REQUIRED: DefInput[] = [
  // Presentation: only these always render empty when missing (required_on_payslip).
  // More keys can be promoted later without changing Digital Form filter logic.
  { key: 'employee_name', section: 'identity', order: 10, category: 'required', required_on_payslip: true },
  { key: 'national_id', section: 'identity', order: 20, category: 'required', required_on_payslip: true },
];

const EXPECTED: DefInput[] = [
  { key: 'employer_name', section: 'employer', order: 25, category: 'expected' },
  { key: 'employer_id', section: 'employer', order: 26, category: 'expected' },
  { key: 'employer_address', section: 'employer', order: 27, category: 'expected' },
  { key: 'employment_start_date', section: 'identity', order: 28, category: 'expected' },
  { key: 'employment_scope', section: 'identity', order: 29, category: 'expected' },
  // Identity block order: name → national_id → employee_number → pay_period
  { key: 'employee_number', section: 'identity', order: 22, category: 'expected' },
  {
    key: 'pay_period',
    section: 'identity',
    order: 24,
    category: 'expected',
    required_on_payslip: true,
  },
  { key: 'base_salary', section: 'earnings', order: 40, category: 'expected' },
  { key: 'salary_calculation_basis', section: 'earnings', order: 45, category: 'expected' },
  { key: 'gross_salary', section: 'earnings', order: 50, category: 'expected' },
  { key: 'income_tax', section: 'deductions', order: 60, category: 'expected' },
  { key: 'national_insurance', section: 'deductions', order: 70, category: 'expected' },
  { key: 'total_deductions', section: 'deductions', order: 80, category: 'expected' },
  { key: 'net_salary', section: 'payment', order: 90, category: 'expected' },
  { key: 'amount_paid', section: 'payment', order: 95, category: 'expected' },
  { key: 'payment_method', section: 'payment', order: 100, category: 'expected' },
  { key: 'minimum_wage_monthly', section: 'other', order: 105, category: 'expected' },
  { key: 'minimum_wage_hourly', section: 'other', order: 106, category: 'expected' },
  { key: 'employee_id', section: 'identity', order: 115, category: 'expected' },
  { key: 'employment_type', section: 'identity', order: 120, category: 'expected' },
  { key: 'seniority_years', section: 'identity', order: 125, category: 'expected' },
  { key: 'department', section: 'identity', order: 130, category: 'expected' },
  { key: 'hourly_rate', section: 'earnings', order: 140, category: 'expected' },
  { key: 'regular_hours', section: 'earnings', order: 150, category: 'expected' },
  { key: 'overtime_hours', section: 'earnings', order: 160, category: 'expected' },
  { key: 'travel_expenses', section: 'earnings', order: 170, category: 'expected' },
  { key: 'health_tax', section: 'deductions', order: 180, category: 'expected' },
  { key: 'pension_employee', section: 'deductions', order: 190, category: 'expected' },
  { key: 'pension_employer', section: 'deductions', order: 200, category: 'expected' },
  { key: 'severance', section: 'deductions', order: 210, category: 'expected' },
  { key: 'training_fund', section: 'deductions', order: 220, category: 'expected' },
  { key: 'bank_name', section: 'payment', order: 230, category: 'expected' },
  { key: 'bank_branch', section: 'payment', order: 231, category: 'expected' },
  { key: 'bank_account', section: 'payment', order: 232, category: 'expected' },
  { key: 'vacation_balance', section: 'other', order: 240, category: 'expected' },
  { key: 'sick_leave_balance', section: 'other', order: 250, category: 'expected' },
  { key: 'messages', section: 'other', order: 260, category: 'expected' },
];

/** Keys known to the backend PAYSLIP_FIELD_KEYS tuple (must stay aligned). */
export const BACKEND_PAYSLIP_FIELD_KEYS = [
  'employee_name',
  'employee_id',
  'employee_number',
  'pay_period',
  'employment_type',
  'department',
  'hourly_rate',
  'base_salary',
  'travel_expenses',
  'regular_hours',
  'overtime_hours',
  'gross_salary',
  'income_tax',
  'national_insurance',
  'health_tax',
  'pension_employee',
  'pension_employer',
  'severance',
  'training_fund',
  'net_salary',
  'vacation_balance',
  'sick_leave_balance',
  'payment_method',
  'messages',
] as const;

/** Mirror of backend PAYSLIP_CANONICAL_EXTRA_KEYS. */
export const BACKEND_PAYSLIP_CANONICAL_EXTRA_KEYS = [
  'national_id',
  'total_deductions',
  'employer_name',
  'employer_id',
  'employer_address',
  'employment_start_date',
  'seniority_years',
  'employment_scope',
  'salary_calculation_basis',
  'amount_paid',
  'bank_name',
  'bank_branch',
  'bank_account',
  'minimum_wage_monthly',
  'minimum_wage_hourly',
] as const;

function buildDefinition(input: DefInput): PayslipFieldDefinition {
  return {
    canonical_key: input.key,
    label_i18n_key: `payroll.fields.${input.key}`,
    section: input.section,
    display_order: input.order,
    requirement_category: input.category,
    editable: true,
    required_on_payslip: Boolean(input.required_on_payslip),
    required_for_persistence:
      input.required_for_persistence ?? PERSISTENCE_CRITICAL.has(input.key),
  };
}

function buildRegistry(): Record<string, PayslipFieldDefinition> {
  const defs: Record<string, PayslipFieldDefinition> = {};
  for (const item of [...REQUIRED, ...EXPECTED]) {
    defs[item.key] = buildDefinition(item);
  }
  let order = 900;
  for (const key of [...BACKEND_PAYSLIP_FIELD_KEYS, ...BACKEND_PAYSLIP_CANONICAL_EXTRA_KEYS]) {
    if (defs[key]) continue;
    defs[key] = buildDefinition({
      key,
      section: 'other',
      order,
      category: 'other',
    });
    order += 10;
  }
  return defs;
}

export const PAYSLIP_FIELD_REGISTRY = buildRegistry();

/**
 * Keys that may carry National ID for display/compat.
 * Prefer national_id. Legacy employee_id only counts when it looks like an Israeli ID.
 */
export const NATIONAL_ID_FIELD_KEYS = ['national_id'] as const;

export function getPayslipFieldDefinition(key: string): PayslipFieldDefinition | null {
  const normalized = key.trim();
  if (!normalized) return null;
  return PAYSLIP_FIELD_REGISTRY[normalized] ?? null;
}

export function requirementCategoryForKey(key: string): FieldRequirementCategory {
  return getPayslipFieldDefinition(key)?.requirement_category ?? 'other';
}

export function requiredOnPayslipKeys(): string[] {
  return Object.values(PAYSLIP_FIELD_REGISTRY)
    .filter((item) => item.required_on_payslip)
    .sort((a, b) => a.display_order - b.display_order)
    .map((item) => item.canonical_key);
}

export function displayOrderForKey(key: string): number {
  return getPayslipFieldDefinition(key)?.display_order ?? 10_000;
}

/** True when value looks like Israeli National ID digits (legacy employee_id compat). */
export function looksLikeNationalIdDigits(value: unknown): boolean {
  if (value == null) return false;
  const digits = String(value).replace(/\D/g, '');
  return digits.length === 9;
}

export function registrySnapshotForSync(): Record<
  string,
  {
    requirement_category: FieldRequirementCategory;
    section: PayslipFieldSection;
    display_order: number;
    required_on_payslip: boolean;
    required_for_persistence: boolean;
  }
> {
  const out: ReturnType<typeof registrySnapshotForSync> = {};
  for (const [key, item] of Object.entries(PAYSLIP_FIELD_REGISTRY).sort(([a], [b]) =>
    a.localeCompare(b),
  )) {
    out[key] = {
      requirement_category: item.requirement_category,
      section: item.section,
      display_order: item.display_order,
      required_on_payslip: item.required_on_payslip,
      required_for_persistence: item.required_for_persistence,
    };
  }
  return out;
}
